File: backend/app/schemas/document_delivery.py
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.schemas.base import UTCModel

DeliveryEntity = Literal['quote', 'purchase_order']
DeliveryStatus = Literal['prepared', 'sending', 'accepted', 'failed', 'unknown']


class DocumentDeliveryPreview(BaseModel):
    entity_type: DeliveryEntity
    entity_id: int = Field(gt=0)
    new_attempt: bool = False


class DocumentDeliverySend(BaseModel):
    expected_version: int = Field(ge=1)
    request_key: str = Field(min_length=8, max_length=100, pattern=r'^[A-Za-z0-9_-]+$')
    recipient: EmailStr = Field(max_length=320)
    subject: str = Field(min_length=1, max_length=200)
    body: str = Field(min_length=1, max_length=10000)

    @field_validator('subject')
    @classmethod
    def single_line_subject(cls, value):
        if '\r' in value or '\n' in value or not value.strip():
            raise ValueError('Subject must be one nonempty line')
        return value.strip()


class DocumentDeliveryReconcile(BaseModel):
    expected_version: int = Field(ge=1)
    outcome: Literal['accepted', 'failed']
    verification_note: str = Field(min_length=10, max_length=1000)

    @field_validator('verification_note')
    @classmethod
    def meaningful_note(cls, value):
        if len(value.strip()) < 10:
            raise ValueError('Describe how the mail-server outcome was verified')
        return value.strip()


class DocumentDeliveryResponse(UTCModel):
    id: int
    entity_type: DeliveryEntity
    entity_id: int
    document_number: str
    issue_date: Optional[date] = None
    recipient: str
    subject: str
    body: str
    attachment_name: str
    attachment_sha256: str
    attachment_size: int
    status: DeliveryStatus
    status_detail: Optional[str]
    version: int
    provider_message_id: Optional[str]
    created_at: datetime
    attempted_at: Optional[datetime]
    accepted_at: Optional[datetime]
    send_available: bool
    unavailable_reason: Optional[str]
    delivered: None = None  # SMTP acceptance cannot establish recipient delivery.
    replayed: bool = False
    manually_verified: bool = False
    verified_at: Optional[datetime] = None
    verification_note: Optional[str] = None

File: frontend/src/types/documentDelivery.ts
export type DeliveryEntity = 'quote' | 'purchase_order';
export type DeliveryStatus = 'prepared' | 'sending' | 'accepted' | 'failed' | 'unknown';
export interface DocumentDelivery {
  id: number;
  entity_type: DeliveryEntity;
  entity_id: number;
  document_number: string;
  issue_date?: string | null;
  recipient: string;
  subject: string;
  body: string;
  attachment_name: string;
  attachment_sha256: string;
  attachment_size: number;
  status: DeliveryStatus;
  status_detail: string | null;
  version: number;
  provider_message_id: string | null;
  created_at: string;
  attempted_at: string | null;
  accepted_at: string | null;
  send_available: boolean;
  unavailable_reason: string | null;
  delivered: null;
  replayed: boolean;
  manually_verified: boolean;
  verified_at: string | null;
  verification_note: string | null;
}
export interface DocumentDeliverySend {
  expected_version: number;
  request_key: string;
  recipient: string;
  subject: string;
  body: string;
}

export interface DocumentDeliveryReconcile {
  expected_version: number;
  outcome: 'accepted' | 'failed';
  verification_note: string;
}
